### PR TITLE
Form-Data and multi-part request feature added

### PR DESCRIPTION
**Pull request by vidya-hub**
### **Multi-Part Request Feature  #92,#91  **

- Added form-data option in request body content type.
- now we can select text or file type in form-data fields and send multi-part Request.
- Form-Data can be **Saved** in Local DB and can **Load** from the saved data.
- Added multipart Code generation feature for all supported languages in the Code Gen DropDown

### CODE REVIEW

File: lib/consts.dart [hunk 1]
```diff
@@ -1,9 +1,10 @@
-import 'dart:io';
 import 'dart:convert';
+import 'dart:io';
+
+import 'package:davi/davi.dart';
 import 'package:flutter/foundation.dart';
 import 'package:flutter/material.dart';
 import 'package:google_fonts/google_fonts.dart';
-import 'package:davi/davi.dart';
```
Comment (ashitaprasad) [inline lib/consts.dart:8]: Feedback in regards to import across all the files.
Please revert all import re-arrangements that have been done as they are introducing non-essential code changes. 
Only add new import lines or modify existing lines.

File: lib/consts.dart [hunk 2]
```diff
@@ -294,6 +298,15 @@ const kSubTypeDefaultViewOptions = 'all';
 const kContentTypeMap = {
   ContentType.json: "$kTypeApplication/$kSubTypeJson",
   ContentType.text: "$kTypeText/$kSubTypePlain",
+  ContentType.formdata: "multipart/form-data",
+};
+const kFormDataTypeMap = {
```
Comment (ashitaprasad) [inline lib/consts.dart:303]: kFormDataTypeMap & kMapFormDataType can be avoided if `FormDataType` is structured as `CodegenLanguage`

File: lib/utils/convert_utils.dart [hunk 3]
```diff
@@ -1,7 +1,11 @@
-import 'dart:typed_data';
 import 'dart:convert';
-import '../models/models.dart';
+import 'dart:typed_data';
+
+import 'package:apidash/models/form_data_model.dart';
```
Comment (ashitaprasad) [inline lib/utils/convert_utils.dart:4]: it should already be imported via '../models/models.dart'

File: lib/services/http_service.dart [hunk 4]
```diff
@@ -63,3 +65,59 @@ Future<(http.Response?, Duration?, String?)> request(
     return (null, null, uriRec.$2);
   }
 }
+
+Future<(http.Response?, Duration?, String?)> multiPartRequest(
```
Comment (ashitaprasad) [inline lib/services/http_service.dart:69]: Can the existing function be modified to support multipart using a multipart = true argument

File: lib/services/http_service.dart [hunk 5]
```diff
@@ -63,3 +65,59 @@ Future<(http.Response?, Duration?, String?)> request(
...
+        requestUrl,
+      );
+      for (FormDataModel formData in (requestModel.formDataList ?? [])) {
+        if (formData.type == FormDataType.text) {
+          request.fields.addAll({formData.name: formData.value});
+        } else {
+          request.files.add(
+            await http.MultipartFile.fromPath(
+              formData.name,
+              formData.value,
```
Comment (ashitaprasad) [inline lib/services/http_service.dart:105]: Why the data type of value in FormDataModel dynamic if it is aString value or String path.

**Comment by ashitaprasad** [comment]
Thanks for the PR @vidya-hub 
Apart from the comments mentioned  above. PFA a few more observations.

Screenshot 2023-12-22 at 9 23 12 AM

There are overflow issues. It should be along the lines of params and headers which have an empty default row in the beginning and the user is able to clear the first row (which gets replaced by an empty row) if it is the only row. Please play around with headers & params to better understand the behaviour.

Also, the color scheme of the file picker widget (purple) is not following the color theme of API Dash.

**Comment by vidya-hub** [comment]
Hi @ashitaprasad 

Thanks for your comments and observations.
Sure I will work on those things and once it is done I will push the code changes.

Thank you.

File: lib/consts.dart [hunk 6]
```diff
@@ -48,6 +48,10 @@ const kHintOpacity = 0.6;
 const kForegroundOpacity = 0.05;
 
 const kTextStyleButton = TextStyle(fontWeight: FontWeight.bold);
+const kFormDataButton = TextStyle(
```
Comment (ashitaprasad) [inline lib/consts.dart:51]: `kFormDataButton` -> `kFormDataButtonLabelTextStyle`

**Comment by ashitaprasad** [comment]
Hi @vidya-hub 
I have reviewed the PR. There are still pending issues that you need to fix that I pointed out in the earlier review.

**Comment by vidya-hub** [comment]
Hey @ashitaprasad 

Thank you 

Can you tell me which issues I've missed so that i can give a fix for that

File: lib/consts.dart (hunk 1 above)
Comment (ashitaprasad) [reply lib/consts.dart]: Not resolved @vidya-hub

**Comment by ashitaprasad** [comment]
Sure @vidya-hub 

This [comment](https://github.com/foss42/apidash/pull/101/files#r1434689867) and this [comment](https://github.com/foss42/apidash/pull/101#issuecomment-1867210865)

File: lib/providers/collection_providers.dart [hunk 7]
```diff
@@ -153,10 +157,13 @@ class CollectionStateNotifier
...
     final defaultUriScheme =
         ref.read(settingsProvider.select((value) => value.defaultUriScheme));
-
     RequestModel requestModel = state![id]!;
-    var responseRec =
-        await request(requestModel, defaultUriScheme: defaultUriScheme);
+    (http.Response?, Duration?, String?)? responseRec = await request(
+      requestModel,
+      defaultUriScheme: defaultUriScheme,
+      isMultiPartRequest: requestModel.formDataList != null &&
```
Comment (ashitaprasad) [inline lib/providers/collection_providers.dart:164]: `requestBodyContentType` is the correct way to determine the value of `isMultiPartRequest`.

File: lib/screens/home_page/collection_pane.dart [hunk 8]
```diff
@@ -108,36 +108,40 @@ class _RequestListState extends ConsumerState<RequestList> {
       controller: controller,
```
Comment (ashitaprasad) [inline lib/screens/home_page/collection_pane.dart:108]: Revert this change which has nothing to do with formdata & we already show a full page message in details pane when there is no request.

File: lib/screens/home_page/editor_pane/details_card/request_pane/request_body.dart [hunk 9]
```diff
@@ -3,6 +3,7 @@ import 'package:flutter_riverpod/flutter_riverpod.dart';
 import 'package:apidash/providers/providers.dart';
 import 'package:apidash/widgets/widgets.dart';
 import 'package:apidash/consts.dart';
+import 'package:apidash/widgets/form_data_widget.dart';
```
Comment (ashitaprasad) [inline lib/screens/home_page/editor_pane/details_card/request_pane/request_body.dart:6]: Add `form_data_widget.dart` to `widgets.dart` which is already imported.

File: lib/screens/home_page/editor_pane/details_card/request_pane/request_body.dart [hunk 10]
```diff
@@ -12,12 +13,21 @@ class EditRequestBody extends ConsumerStatefulWidget {
 }
 
 class _EditRequestBodyState extends ConsumerState<EditRequestBody> {
+  @override
```
Comment (ashitaprasad) [inline lib/screens/home_page/editor_pane/details_card/request_pane/request_body.dart:16]: Remove as you are not doing anything inside `initState()`

**Comment by ashitaprasad** [comment]
Is this ready for review @vidya-hub?

**Comment by vidya-hub** [comment]
Yes @ashitaprasad

**Comment by ashitaprasad** [comment]
@vidya-hub I have made some changes to the PR.
Can you go through it once and test it out.

**Comment by vidya-hub** [comment]
Sure @ashitaprasad

**Comment by vidya-hub** [comment]
@ashitaprasad 

I've gone through the changes and tested it everything is working fine only related to form-data.
Thanks.

**Comment by ashitaprasad** [comment]
Hi @vidya-hub,
I have fixed some tests. 
But, currently js & nodejs codegen tests are failing. This might be because of the lack of check in the codegens for the presence of any form data.
Kindly, run `flutter test` to see the detailed test results and kindly fix the codegens.
Thanks!

**Comment by vidya-hub** [comment]
Thank you @ashitaprasad 

I will check and update.

**Comment by vidya-hub** [comment]
image


Hi @ashitaprasad 
I've fixed all flutter-test-related issues and all tests are passed can you please check once.

Thanks!

**Comment by ashitaprasad** [comment]
Sure @vidya-hub 👍

**Comment by ashitaprasad** [comment]
@vidya-hub Since you are not adding any new test cases for multipart forms, please make changes only to the codegen codes revert any changes made to the expected codes (even spacings) given in the test files. 
These expected codes are golden tests that have already been run, verified and formatted properly.

**Comment by vidya-hub** [comment]
Sure @ashitaprasad

**Comment by vidya-hub** [comment]
Hi @ashitaprasad 

test/models/request_model_test.dart 

can you please check this file in PR once

**Comment by ashitaprasad** [comment]
@vidya-hub `test/models/request_model_test.dart` will only change as the model has been changed.
I was talking about the codegen test cases that contain the expected code.

**Comment by vidya-hub** [comment]
Thank you,
Understood @ashitaprasad.

I've pushed the changes,
can you please check once .

**Comment by ashitaprasad** [comment]
Sure @vidya-hub 👍

**Comment by ashitaprasad** [comment]
@vidya-hub There was a minor issue with python codegen which I have fixed.
LGTM. It is ready for merge 🚀
The only thing remaining is adding codegen tests for this feature which we can add in a separate PR once I open an issue & assign it to you after figuring out a good public endpoint.
Thanks for the awesome PR 🙌

**Comment by vidya-hub** [comment]
Thanks a lot @ashitaprasad 

Sure I am working on it already.

Thank you for your comments and suggestions.
I've learnt a lot from you by implementing this feature.

**Comment by ashitaprasad** [comment]
@vidya-hub 
It was a blast working with you on this valuable PR! 🎉
Definitely looking forward to more contributions from you 🤘
Don't forget to share this wonderful open source achievement on your socials (Twitter, LinkedIn). 🥳

**Comment by vidya-hub** [comment]
Thank you @ashitaprasad.

Sure 😊
